npm/src/log.ts

```diff
@@ -63,12 +69,46 @@ export const log = {
     // Utility function to ensure exhaustive type checking. See https://stackoverflow.com/a/39419171
     log.error("Exhaustive type checking didn't account for: %o", val);
   },
+  /**
+   * @param level - A number indicating severity: 1 = Error, 2 = Warn, 3 = Info, 4 = Debug, 5 = Trace
+   * @param target - The area or component sending the messsage, e.g. "parser" (useful for filtering)
+   * @param args - The format string and args to log, e.g. ["Index of %s is %i", str, index]
+   */
+  logWithLevel(level: number, target: string, ...args: any) {
+    // Convert to a format string containing the target (if present)
+    const [, ...trailingArgs] = args; // All but first element of args
+    const outArgs = [`[%s] ${args[0]}`, target || "", ...trailingArgs];
```

Comment:
@minestarks this was the kind of thing/reason I was asking about the VS Code logging. Will this break or be weird then when redirected containing a format string?

Reply:
Ah. Yes, it will look weird with my current code, but I just realized VS Code's output channel API provides `console.log`/`console.error`-like functions. I'll update my code to use those instead and it should be fine.

https://code.visualstudio.com/api/references/vscode-api#LogOutputChannel

Reply:
Done - it cleaned up really nicely https://github.com/microsoft/qsharp/blob/e3816a5f88450e446f49ad0a1ec8c2a6cebdea8a/vscode/src/extension.ts#L54